### DISCUSSION TITLE
Support embedded il instructions

### DISCUSSION THREAD

**Discussion by muhamad** [dotnet/csharplang#927] [General]
This is a proposal to add new keyword to allow writing inline IL code directly inside C# code for example

```csharp
using System;

public void WriteMessage()
{
    var message = "a message to console";

    il
    {
        ldlocal.0
        call void Console::WriteLine(string)
    };
}
```

**Comment by HaloFour** [comment]
If I recall it was considered quite an effort to try to add IL to the C# specification, including all of the glue syntax that would be required for IL to support C# constructs and vice versa.  The team instead is looking to expose "compiler intrinsics", or special methods that would be converted by the compiler into IL operations, in order to represent special IL operations that cannot be generated easily by normal C# code.  https://github.com/dotnet/roslyn/issues/11475

**Comment by HaloFour** [comment]
I couldn't find a previous request for embedded IL on the [Roslyn](https://github.com/dotnet/roslyn/) repo or [CodePlex](http://roslyn.codeplex.com/discussions/topics/5998/c-language-design).  Could be that my Search-Fu is failing me.  Anyway, the issue on compiler intrinsics addresses the concept but it's good to have this issue as a place to discuss the concept.

**Comment by jnm2** [comment]
@HaloFour https://github.com/dotnet/roslyn/issues/13735


Edit: Oh hey https://github.com/dotnet/csharplang/issues/226

**Comment by HaloFour** [comment]
Figures, I was only looking on Roslyn.

Anyway, #428 is more like this issue whereas #226 is much more open ended.

**Comment by jnm2** [comment]
Fwiw I searched dotnet/roslyn `is:issue asm il` and skimmed the 11 titles that showed up.

**Comment by muhamad** [comment]
In MSDN, they have example of using Dynamic binding before DLR and after
```csharp
/*before*/ Scriptobj.SetProperty("Count", ((int)GetProperty("Count")) + 1);
/*after*/  scriptobj.Count += 1;
```
so if it's going to be so much effort to add inline assembler why just not treat it as a dynamic language - which it is from JIT point-of-view - and embed it in the output and let JIT do the checking.

and let me say that it is allowed for us to add dynamic IL using `ILGenerator` and if we add invalid instruction it get caught at runtime and and exception thrown which we are responsible for emitting invalid instruction.

**Comment by HaloFour** [comment]
@muh00mad 

> so if it's going to be so much effort to add inline assembler why just not treat it as a dynamic language - which it is from JIT point-of-view - and embed it in the output and let JIT do the checking.

The compiler would still be required to take the text of the IL instructions and compile them into the corresponding IL byte code.  That has to happen at compile time.  The CLR/JIT has no concept what `ldlocal.0` is, it only knows `0x06` which is the single byte instruction that it compiles to.  The only way to do that within the compiler is to extend the compiler to understand IL syntax.

And then there is the issue of integration.  What does `ldlocal.0` mean embedded within a C# method?  The locals emitted by the C# compiler don't have to be in any particular order and the compiler is free to both optimize out declared locals as well as introduce additional unnamed locals.  This is why in IL you assign slot numbers to each variable.  But as an embedded language you'd need the C# compiler and the IL compiler to work together using a dialect so that you can reference the local by name.

**Comment by svick** [comment]
Why? When would you want to use this (assuming the https://github.com/dotnet/roslyn/issues/11475 IL intrinsics existed)?

**Comment by mattwar** [comment]
This feature suggestion makes me IL. 😆

**Comment by muhamad** [comment]
> The compiler would still be required to take the text of the IL instructions and compile them into the corresponding IL byte code. That has to happen at compile time. The CLR/JIT has no concept what ldlocal.0 is, it only knows 0x06 which is the single byte instruction that it compiles to. The only way to do that within the compiler is to extend the compiler to understand IL syntax.

AFAIK the work already been done [here](http://xoofx.com/blog/2016/05/25/inline-il-asm-in-csharp-with-roslyn/) and all it needs is a some - minor - effort to add into the compiler source.

> Why? When would you want to use this (assuming the dotnet/roslyn#11475 IL intrinsics existed)?

Some project I'm working on require fast alternative to `PropertInfo.SetValue` with simple interface for ValueTypes and ReferenceType so the function was `SetValue(object instance, string name, object Value)`
which works fine for reference types but for value type it didn't work at all for obvious reasons so here is the example I implemented to solve the problem using `DynamicMethod`, this example works fine even if `SampleType` was reference type

```csharp
struct SampleType
{
    public int Number { get; set; }
}

class Program
{
    public static void Main()
    {
        var type = default(SampleType);
        object o = type;
        SetValue(o, nameof(SampleType.Number), 10)
    }

    public static void SetValue(object instance, string name, object value)
    {
        if (name == "Number")
        {
            // invalid in c#
            //((SampleType)instance).Number = (int)value;

            // valid in c#, the modified object is not the target object but the local one
            //SampleType sampleType = (SampleType)instance;
            //sampleType.Number = (int)value;

            il
            {
                ldarg.1
                unbox SampleType
                ldarg.3
                unbox.any [mscorlib]System.Int32
                call instance void SampleType::set_Number(int32)
            }
            return;
        }
        throw new Exception("name not found");
    }
}
```
The compiler intrinsics is not confirmed yet and if it is going to be added it will be some functions so the IL code above will be something like
```csharp
using Microsoft.CSharp.Intrinsics;

    CompilerIntrinsics.LdArg(1); // CompilerIntrinsics.LdArg(int index)
    CompilerIntrinsics.Unbox(typeof(SampleType)); // CompilerIntrinsics.Unbox(type)
    CompilerIntrinsics.LdArg(3); // CompilerIntrinsics.LdArg(int index)
    CompilerIntrinsics.UnboxAny(typeof(SampleType)); // CompilerIntrinsics.UnboxAny(type)
    CompilerIntrinsics.Call(typeof(SampleType).GetProperty(nameof(Range.Number).GetSetMethod); // CompilerIntrinsics.Call(MethodInfo metthod)
```
which is more effective from end-user point-of-view?

**Comment by mattwar** [comment]
I think you both want ldarg.0 and ldarg.2

**Comment by bondsbw** [comment]
I'd still like to see either multi-language projects or language extensibility.

Compiler intrinsics will also be welcome, but it solves a different set of problems.

**Comment by HaloFour** [comment]
@muh00mad 

> AFAIK the work already been done here and all it needs is a some - minor - effort to add into the compiler source.

That project was already referenced, and no, it would not take minor work.  It would take massive complicated work in order to integrate with the compiler, IDE, debugger, E&C, refactoring, etc.  It's not sufficient to hack it in there.

> which is more effective from end-user point-of-view?

For something that should be exceptionally rarely used?  Probably the one which has any likelihood of getting implemented and would actually get the job done.

**Comment by svick** [comment]
@muh00mad 

> The compiler intrinsics is not confirmed yet and if it is going to be added it will be some functions so the IL code above will be something like

No, it won't. There is nothing in the proposal about supporting "common" IL instructions like `ldarg` or `call`. So it wouldn't work for this use case.

**Comment by svick** [comment]
@muh00mad 

I think that use case is an argument to add an `Unbox` method, which uses `ref` return to avoid the copy, either as an intrinsic or similar to methods on the `Unsafe` type (which are just simple methods implemented in IL).

With that, your code could look like this:

```c#
public static void SetValue(object instance, string name, object value)
{
    if (name == "Number")
    {
        ref SampleType sampleType = ref Unbox<SampleType>(instance);

        sampleType.Number = (int)value;

        return;
    }
    throw new Exception("name not found");
}
```

The implementation of `Unbox`:

```il
.method public hidebysig static !!T& Unbox<valuetype T>(object o) cil managed aggressiveinlining
{
    .maxstack 1
    ldarg.0
    unbox !!T
    ret
}
```

**Comment by fanoI** [comment]
I think that if this thing will be made there should be the possibility from the IL to access the C# variables directly (that is it will not the "real" IL) as does - for example - GCC with inline assembly:

```cs
using System;

public void WriteMessage()
{
    var message = "a message to console";

    il
    {
        call void Console::WriteLine(message)
    };
}
```

**Comment by muhamad** [comment]
@HaloFour 
> That project was already referenced, and no, it would not take minor work. It would take massive complicated work in order to integrate with the compiler, IDE, debugger, E&C, refactoring, etc. It's not sufficient to hack it in there.

OK, lets break it down a little,

First; if you compile c# code for debug the generated MSIL can not be seen or debugged from IDE, in fact if you go to the assembly window all you can see is the assembly language produced by JIT.

Second; you can not debug dynamic languages AFAIK. Actually the C# 7.1 compiler produce dynamic bounding code for the following snippets in debug and release mode just because i used `dynamic` keyword even though the target code is a c# code.(1)

```csharp
struct SampleType
{
    private int number;

    public int Number
    {
        get => number;
        set => number = value;
    }
}

class Program
{
    public static void Main()
    {
        dynamic st = new SampleType();

        st.Number = 10;
        int n = st.Number;
    }
}
```
Third; you can not refactor a dynamic-based code, yet you can E&C it.

What I'm saying here is dynamic code was found to solve certain problems that **had a solution** or work-around before dynamic exists but found to make code even easier and to solve more problems using unified model, same for MSIL - from my POV - which solve common problems, maybe not for many people but the same can be said for dynamic code.

Implementing MSIL don't need to have refactor which not exist for dynamic code nor E&C because the IDE itself don't support debugging IL code. some features can be added and grow later IMO.


(1) i know that is the right behavior as the documentation says but my point here is the compiler acted on the internal value type as if it not exist in passed the execution to DLR which it did the work.

**Comment by HaloFour** [comment]
@muh00mad 

>  First; if you compile c# code for debug the generated MSIL can not be seen or debugged from IDE, in fact if you go to the assembly window all you can see is the assembly language produced by JIT.

That's because the compiler doesn't generate debugging symbols for two languages.  This would be one language, and yes, the compiler would be *expected* to produce debugging symbols here.

> Second; you can not debug dynamic languages AFAIK.

You absolutely can.  I do it all the time.  You can step through the C# code that is using `dynamic` types, and you can debug into the dynamic invocations themselves.

> Third; you can not refactor a dynamic-based code

Because dynamic code has no known tokens at compile time.  That's the entire point of dynamic code.  IL is *not* a dynamic language.  All of the types, members, etc. are (and must) be statically resolved at compile time.

> Implementing MSIL don't need to have refactor

If that embedded IL can reference C# tokens then yes, it absolutely must support refactoring.

> nor E&C because the IDE itself don't support debugging IL code.

The IDE supports this just fine.  You need a debugging symbol database which would correlate the binary back to the source file.  The C# compiler doesn't generate IL source nor additional debugging symbols.  There's no technical reason why the C# compiler can't do this, there's just fairly little reason to do so.  But `ilasm.exe` does, and you can use that to debug IL just fine.

>  some features can be added and grow later IMO.

The C# compiler team is not in the habit of implementing features without proper first-class support throughout their tools.  This is the one reason why source generators have yet to make it into the language; the team refuses to implement the feature until they figure out how it works with the tools.  They would certainly not consider adding something like embedded IL without tooling support in the IDE.  It would be impossible to modify the implementation after the fact in order to support tooling as the Roslyn compiler is not a black box and it has a public API which is expected to be stable.

And that still involves taking an entire language, which is IL, and shoving it into the C# spec.  The C# spec is already huge and complicated.

My suggestion is to take those use cases where you've felt that you've had to break out IL and to request helper intrinsic methods in the BCL.  The specific case you mentioned, mutating a boxed struct, can easily be solved with such a method.  And it doesn't require this massive effort to accomplish.

**Comment by svick** [comment]
@muh00mad 

> What I'm saying here is dynamic code was found to solve certain problems that had a solution or work-around before dynamic exists but found to make code even easier and to solve more problems using unified model, same for MSIL - from my POV - which solve common problems, maybe not for many people but the same can be said for dynamic code.

What are those "common problems"? As far as I can see, so far you've shown just one problem, which could be solved by a library method. If you want a feature as big as this added to the language, you will be need a really good justification.

**Comment by AdamSpeight2008** [comment]
How about supporting it at the method definition level?
Example
```c#
System.IL WriteMessage(string text)
{
  ldlocal.0
  call void System::Console::WriteLine(text)
}
```

**Comment by IS4Code** [comment]
Well, at least on the method level, allowing CIL code would help creating "intrinsic" methods using low-level concepts. The compiler would also have the option of not emitting direct IL but use *DynamicMethod* and *ILGenerator* instead.

**Comment by The-Futurist** [comment]
@HaloFour @svick - Is there any chance this might get more attention? Perhaps the grammar could add a new keyword `embed `along with some way to specify what to embed:

```
public void GetInfo (Type Type)
{
   embed(Embedding.IL)
   {
   ldlocal.0
   etc etc etc.
   }
}

```
Since IL is truly platform independent I see no huge reason to not support some form of support for embedding IL.

Another thing I've often wanted is making IL a first class .Net language fully supported in Visual Studio so I could have some C# solution but with some dedicated IL project in which I can code IL directly (.il files) along with ability to debug and so on.

In the past I've had to resort to dynamic code generation which by comparison is very cumbersome and hard to debug.

It would also be great to be able to debug at the IL level but we cannot, for example if we debug code and have no source for it then we are faced with debugging raw machine assembler despite the fact that the IL exists (and has been converted to IL) it would be great to debug optimized code for which we have no source - at the IL level.

In fact adding support for full IL projects might eliminate any need to add IL embedding to C# at all thus keeping C# pure.

**Comment by svick** [comment]
@Korporal 

> Is there any chance this might get more attention?

If you're asking what would be the next step to make this happen, then that would require [a "champion" from the Language Design Team](https://github.com/dotnet/csharplang/#design-process).

> I see no huge reason to not support some form of support for embedding IL

It's the other way around: each feature (especially one as complicated as this one) needs a really good reason to exist. It's not good enough that you don't see a huge reason for it not to exist.

As I've asked on this issue before: what are the use cases for this proposal? What makes this proposal better than cheaper alternatives like IL intrinsics and library methods implemented in IL?

**Comment by The-Futurist** [comment]
@svick 

> As I've asked on this issue before: what are the use cases for this proposal? What makes this proposal better than cheaper alternatives like IL intrinsics and library methods implemented in IL?

and I had written

> Another thing I've often wanted is making IL a first class .Net language fully supported in Visual Studio so I could have some C# solution but with some dedicated IL project in which I can code IL directly (.il files) along with ability to debug and so on.

Which we seem to agree, is better in many ways than adding embedding support to C#, but where to discuss this idea about making IL a first class Visual Studio language, project type?

Now you mentioned "library methods implemented in IL" and how do these get created? what Visual Studio support is the for translating IL into fully fledged assembly - COFF - DLL files?

**Comment by tannergooding** [comment]
> but where to discuss this idea about making IL a first class Visual Studio language, project type?

@Korporal, I would recommend starting on the https://github.com/dotnet/project-system repository, as well as potentially opening an issue on https://github.com/dotnet/sdk.

Project System would be the likely place to add VS support (or forward it to the relevant team).
SDK would be the likely place to get command line support.

**Comment by orthoxerox** [comment]
I think I've already seen a VS extension that lets you write projects in IL.

**Comment by The-Futurist** [comment]
@orthoxerox - Yes but was pretty feeble if I recall and in a rather beta state.

**Comment by svick** [comment]
@Korporal 

> Now you mentioned "library methods implemented in IL" and how do these get created? what Visual Studio support is the for translating IL into fully fledged assembly - COFF - DLL files?

I didn't mean that a "regular programmer" is supposed to write IL-based projects. I meant that such a programmer can use projects written in IL as a library (which can be either [part of the framework](https://github.com/dotnet/corefx/blob/95b2163/src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il) or a NuGet library).

And since very few people are going to be writing such libraries, VS support would probably not be worth it.

**Comment by 06needhamt** [comment]
Is there still any interest in this feature? If so, I’d like to have a go at implementing it into Roslyn.

**Reply by atcarter714** [reply]
Yes, I'm definitely interested in it. I do a lot of interop and "make stuff work" hacks that could be more "properly" done with built-in, inline IL blocks. Interested in knowing if you ever tried and what you discovered ...

**Comment by 333fred** [comment]
> Is there still any interest in this feature?

This feature is not championed, so there is no place for it in the schedule currently. See https://github.com/dotnet/csharplang/blob/master/README.md#design-process.

**Comment by AdamSpeight2008** [comment]
@06needhamt Even though the feature isn't championed, if you would like to have a go at implementing the feature. Go ahead and prototype it.

**Comment by 333fred** [comment]
> Go ahead and prototype it.

I want to be clear though: a PR to roslyn with an implementation for a unchampioned, undiscussed feature will be closed. Unless and until the feature is championed by a member of the LDT, the feature is not open for inclusion in the language.

**Comment by CyrusNajmabadi** [comment]
You are, of course, always welcome to do wahtever you want in your own forks.  But Fred is correct, the *explicit* team stance on this sort of thing is that a PR for that would be rejected for not following proper procedure.

**Comment by msedi** [comment]
@06needhamt : As long as there is missing support in C# and some things can only be accomplished by using IL I agree that there needs to be some mechanism. The question is, if you need it integrated in C# or if it's better to have some official and well supported ilproj?

**Comment by tannergooding** [comment]
> to have some official and well supported ilproj?

There is an ilproj SDK that is used by dotnet/runtime. It allows you to easily compile IL into assemblies, but doesn't provide intellisense or VS project system support.

> some things can only be accomplished by using IL I agree that there needs to be some mechanism

The few things that are missing (like function pointers) have championed proposals (generally speaking) and are being worked on.

Other cases can be handled by things like dynamic methods. Dynamic methods or methods explicitly written in pure IL tends to work better as well since the compiler doesn't have to optimize around you inserting arbitrary IL. The same goes for inline asm in C/C++ and is one of the reasons intrinsics are preferred.

**Comment by msedi** [comment]
@tannergooding: I agree with the things that are solved with the championed proposals. I also agree to use intrinsics. The embedded asm {} in C++ was in my opinion only for microoptimizations. But what about math on generics? I can currently only do it in IL. I'm not aware that something has been championed by this.

I needed to write my own assembly that holds basic math on generics, but only for these 6 methods I had to write my own assembly that I could have embedded into my own core assembly if there was a possibility to do so. There are of course arguments around here that an assembly does not matter, but we have around 200 assemblies and in the end it does matter.

If there would be a possibility to add an *.il file to my project that is compiled and added that would be just fine for me. I don't need to embed code directly, only if there's performance drawback in a method call vs. embedded code.

**Comment by tannergooding** [comment]
> But what about math on generics?

Provided it is over a value type, the JIT supports generic specialization which ensures the codegen is optimal.

For example: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNYgGYmpBkIYBvGgzMNT5gPRWGAFQAWvXA2wYMUXsA4YYDZwwQXJIAngwAZtAMGA5+uA7YUGKS2MAMACbOeLgw+MDSUJZmNgwAkhj+uFwA5BVcMDDpjf5c0bGuYBgc2JIMkE1FDADasglJADKp7Nx8AiwAUrwYAOIw9Z5gABQYIWIwEOGbHJJBAOYAlOdogwyjiWKTwNM8/IKLK2syytu7+4e8PEu12o5jMdwmU04Lzm71W62+Oz2B024RObiBN3BD0hM1eCyWcK+W0Rf02yw4vHSlwAuoMhgBZGAxCDpUriSSbRnM1nsgDyYj4QVwLAAgqdTrBcLheAoYPzZrwAF5uXhBc60kHmHRMchIewMEXpdIAHjsAD5NnYGJIHLg0PqoLbzsYbuZeOEGD8kYc7M6ALx+6K/ZEAjCXV1mEya0ExpgAdk9vs2EGAACsYJ1zptNqGsyn05mbS4ANSe3PJtMZsOO3DnDTRmMAXwjDBgkhy/g9XtJvoYAaD3pRaLD5xbUdjsYUiQYkuOFUDAFUuLhsOFBCLcKb7aiIG4LbAPUXnaWlyu16LN3Zt8P9zAPTW6y3QcQE6fV+vNzu3PbzZsDzPVDnR8G1BZsQPMNsO3dT0SWRXt+1gw5jjOcNwMjJ8tQTS080rTNs2QrgLgrAswyLBhSyOE5CJwkjzgfesJwYMDGMgmAxwwswYigCAAHcGHqPjSi4KdJEpfkZBVIIAFEEBUAVVS4TZgMY5jzDAxsgA

You can see in each case where a `typeof(T) == typeof(...)` check exists and `T` is a value type, it optimizes away the dead code (any path where the typeof check would be false). The downsides of this approach are:
* It doesn't work for reference types
* You have to explicitly codify the supported `T`
* The IL gets bigger, so you may need `[MethodImpl(MethodImplOptions.AggressiveInlining)]` on the method
* The code may be less readable due to casts (`(T)(object)result` and `(...)(object)value`) or using `Unsafe.As`

But it works and generates highly optimized code.

**Comment by msedi** [comment]
@tannergooding: Yes, I know this approach and it is far from being optimal. To be honest - and I don't want to sound rude - the fact that it can be done does not mean that it's good. Our code is highly cluttered with the methods described and is also far from being maintainable, since every code path needs to be implemented explicitly. This means that every change means changing all pathes. When you are doing a lot of number crunching this can be annoying like hell ;-)

One can see the result of having no math on generics: Math -> MathF. Even though there might be differences in the internal behaviors, but I think Math.Abs and MathF.Abs is a good example where it would help to have support on generics.

One can see that you only do this once and never touch it again, but it is the time to develop these pieces of code and to maintain them. And as said before in math and physical environments where you need to develop algorithms this could help a lot.

**Comment by IS4Code** [comment]
Powerful JIT can help optimize away certain "pathological" casts, but it doesn't mean the casts are no longer pathological themselves.

I wholeheartedly support this idea, even though specifying it would essentially require to add another language into the specification. In the past, others have also objected that adding arbitrary CIL could make the code horribly wrong or simply invalid, but now with tools like `Unsafe`, programmers have plenty of ways to destroy their lives, if they wish to accept the possibility by having used these tools in the first place. The same could simply happen with embedded CIL, and thanks to the difficulty of the language, not many people would dare to use it anyway. Besides, you can do virtually the same thing with `ILGenerator` already, but that is not supported everywhere and has worse performance.

The remaining question is how much of the syntax should be accepted. I think the most powerful option is the best one - `cil`/`il` should work in a method, class, or global scope, and allow anything that would normally be allowed in the same context in a pure CIL program.

**Comment by HaloFour** [comment]
> I wholeheartedly support this idea, even though specifying it would essentially require to add another language into the specification.

Sounds like a massive amount of effort to slightly better support the vast minority of library writers.

They already have solutions, slightly less than ideal solutions, but given how often this code needs to be written it doesn't sound like it justifies adding IL (and a hybrid form of IL which would allow interop between IL and C# source).  For every one of those use cases there sounds like there are numerous alternative approaches each of which involve many, many magnitudes less effort.

**Comment by tannergooding** [comment]
> One can see the result of having no math on generics: Math -> MathF

This had nothing to do with math/generics but was an overload resolution problem. `System.Math` was created in v1 of the framework and for a large number of functions only supported `double`. Because `int` is implicitly convertible to both `float` and `double`, but prefers `float`, it was a breaking change to add single-precision overloads because it is a silent source breaking change (`double d = Math.Sqrt(5);` would call `Math.Sqrt(float)` rather than `Math.Sqrt(double)`). So we added `MathF` in order to avoid this issue.

You'd likewise have potential problems even if they were generic from day 1, as not all types may be supported and so you wouldn't get a compile time error for things like `Math.Sqrt<Guid>` (you would need to rely on analyzers for that).

> One can see that you only do this once and never touch it again, but it is the time to develop these pieces of code and to maintain them

Yes, it is effectively a write once situation. There are only a limited number of types (the runtime primitives) where you can use IL instructions to perform arithmetic or other logical operations. These can be trivially supported via a small generic helper class and then never need to be touched again.

Any other type ends up being a call to a user-defined operator (including types like `System.Decimal` which is only a C# primitive, not a .NET primitive) and would likely fit into one of the `shapes` proposals that are already being tracked.

**Comment by msedi** [comment]
@tannergooding: I'm curious about the syntax. Let's assume I want to generalize Add and I have primitive times and types with an overloaded + operator. Do I need two different kinds of Add methods:

`T Add<T>(T a, T b) where T : struct`

and another (I don't know the final syntax)

`T Add<T> (T a, T b) where T : shape`

and the CLR emits different pathes or is it then subsumed under one Add method?

**Comment by tannergooding** [comment]
That would depend entirely on how the shapes proposal was implemented, among other factors.

**Comment by cschuchardt88** [comment]
I think this should be more like `C\C++` `__asm { opcodes }` as for Dotnet is should look like  `__ilasm { opcodes }` 
```csharp
    class Program
    {
        static unsafe void Main(string[] args)
        {
            __ilasm
            {
                ldstr "Hello World"
                call System.Console::WriteLine(string)
            }
        }
    }
```

```csharp
    class Program
    {
        static unsafe void Main(string[] args)
        {
           string HelloString = "Hello World";

            __ilasm
            {
                ldstr &HelloString
                call System.Console::WriteLine(string)
            }
        }
    }
```

**Reply by colejohnson66** [reply]
This is already possible with external packages, specifically [InlineIL.Fody](https://github.com/ltrzesniewski/InlineIL.Fody). You can even combine C# and "IL" in the same method.

I would like to see support for this feature, but for now, a solution does exist.

**Reply by cschuchardt88** [reply]
I just want to copy `il` from `ildasm.exe` and have it work. But thank you. Just hate having to call methods.

**Reply by CyrusNajmabadi** [reply]
There is no plan or appetite to do anything here in the language space. If you want to work with IL, there are options available to you.

**Reply by tannergooding** [reply]
Even in languages like C/C++, inline assembly is not a truly standard feature. While the spec does define `asm( string_literal )`, how it actually works is implementation defined and is not universally supported. Using such a feature is also typically considered bad practice and can often lead to worse overall performance due to it causing a general inability for the compiler to actually do its job.

Compilers do not do well when they cannot perform inference over a block of code, when they cannot recognize patterns, do transformations, or when they are otherwise artificially restricted from being able to freely pick the optimal use of CPU/system resources

Manually writing IL is no different and more often that not you're likely de-optimizing your code by writing it, especially if you deviate from the patterns that something like Roslyn would emit (which is largely what the JIT/AOT and other tools are built around recognizing).

There are very few IL features that C# cannot express today and even fewer that are not polyfilled via runtime provided functionality like `System.Runtime.CompilerServices.Unsafe` or similar. The things that can't be expressed are often under-tested and under-optimized by the runtime itself. Because of that, you will almost always get overall better code, performance, stability, readability, maintainability, etc by simply writing standard C# instead of trying to be "clever" and work against the tooling.
